lib/libvfio-user.c

```diff
@@ -540,7 +540,7 @@ handle_dma_unmap(vfu_ctx_t *vfu_ctx, vfu_msg_t *msg,
     assert(msg != NULL);
     assert(region != NULL);
 
-    if (msg->in_size < sizeof(*region)) {
+    if (msg->in_size < sizeof(*region) || msg->in_size < region->argsz) {
```

Comment:
ditto, this is incorrectly requiring an "in" buffer of ->argsz, but that value defines the max *reply* payload

Reply:
ok

Reply:
In this particular case, `region->argsz` must be large enough to accomodate the respone, which is `sizeof(struct vfio_user_bitmap) + region->bitmap->size`, right?

Reply:
It must be at least sizeof(*region): in that case, we need to reply with an out_region, with argsz set to the full required size, as per the spec. That's why you need to have the region struct in the reply as I think I mentioned elsewhere?

So it's clearer as:

```
if (msg->in_size < sizeof(*region) || region->argsz < sizeof(*region))
```

like, for example, handle_device_get_irq_info()

IOW, if we can't even read the in_region, or reply with an out_region, we can't progress.

Then in the bitmap case, we check that ->in_size also has a vfio_user_bitmap, then in the reply when bitmap is requested

```
out_region->argsz = sizeof(*region) + sizeof(struct vfio_user_bitmap) + region->bitmap->size
```

and, *if* it fits into region->argsz, add on the vfio_user_bitmap and the bitmap itself

Does that make sense?

Reply:
handle_device_get_region_info is a better example